base_exception/models/base_exception.py

```diff
@@ -0,0 +1,215 @@
+#       base.exception line (ex rule_group = sale for sale order)
+#  - object: same as order or line, browse_record of the base.exception or
+#    base.exception line
+#  - pool: ORM model pool (i.e. self.pool)
+#  - time: Python time module
+#  - cr: database cursor
+#  - uid: current user id
+#  - context: current context
+""")
+
+
+class BaseException(models.AbstractModel):
+    _name = 'base.exception'
+
+    _order = 'main_exception_id asc'
+
+    main_exception_id = fields.Many2one(
+        'exception.rule',
+        compute='_compute_main_error',
+        string='Main Exception',
+        store=True)
+    rule_group = fields.Selection(
+        [],
+        readonly=True,
+    )
+    exception_ids = fields.Many2many(
+        'exception.rule',
+        string='Exceptions')
+    ignore_exception = fields.Boolean('Ignore Exceptions', copy=False)
+
+    @api.depends('exception_ids', 'ignore_exception')
+    def _compute_main_error(self):
+        for obj in self:
+            if not obj.ignore_exception and obj.exception_ids:
+                obj.main_exception_id = obj.exception_ids[0]
+            else:
+                obj.main_exception_id = False
+
+    @api.multi
+    def _popup_exceptions(self):
+        action = self._get_popup_action()
+        action = action.read()[0]
+        action.update({
+            'context': {
+                'active_id': self.ids[0],
+                'active_ids': self.ids
+            }
+        })
+        return action
+
+    @api.model
+    def _get_popup_action(self):
+        action = self.env.ref('base_exception.action_exception_rule_confirm')
+        return action
+
+    @api.model
+    def _check_exception(self):
+        """
+        This method must be used in a constraint that must be created in the
+        object that inherits for base.exception.
+        for sale :
+        @api.constrains('ignore_exception',)
+        def sale_check_exception(self):
+            ...
+            ...
+            self._check_exception
+        """
+        exception_ids = self.detect_exceptions()
+        if exception_ids:
+            exceptions = self.env['exception.rule'].browse(exception_ids)
+            raise ValidationError('\n'.join(exceptions.mapped('name')))
+
+    @api.multi
+    def test_exceptions(self):
+        """
+        Condition method for the workflow from draft to confirm
+        """
+        if self.detect_exceptions():
+            return False
+        return True
+
+    @api.multi
+    def detect_exceptions(self):
+        """returns the list of exception_ids for all the considered base.exceptions
+        """
+        exception_obj = self.env['exception.rule']
+        model_exceptions = exception_obj.sudo().search(
+            [('model', '=', self._name)])
+        sub_exceptions = exception_obj.sudo().search(
+            [('rule_group', '=', self.rule_group),
+             ('id', 'not in', model_exceptions.ids),
+             ])
```

Comment:
Hi @florian-dacosta , We have need of exceptions applied to model (i.e sale order) and sub exceptions applied to line (i.e sale order line). 
I change the search to make 1 search then filter data on this result.